readme.md
Work in Progress
================

This extension is not working, it's just to visualize an idea instead of a complete package.

Examples
--------

see [RegularWorkspacesTest.php](Tests/Unit/RegularWorkspacesTest.php)

```php
$first = $simulationService->appendWorkspaceElement('tt_content', array('header' => 'First'));
$second = $simulationService->appendWorkspaceElement('tt_content', array('header' => 'Second'));
$third = $simulationService->appendWorkspaceElement('tt_content', array('header' => 'Third'));

// ...

$firstLocalized = $simulationService->localizeLiveElement('tt_content', $first->uid);

// ...

$simulationService->moveElementAfter('tt_content', $first->uid, $second->uid);

// ...

$this->assertSortingOrder(
	'tt_content', 'sorting',
	array(
		$first->placeholderUid,
		$second->placeholderUid,
		$third->placeholderUid,
	)
);
```


ChangeLog
2013-03-03  Oliver Hader  <oliver@typo3.org>

	* [TASK] #29278: Extend test-case for 1n:FF copy localized new structure on workspaces

2013-02-05  Oliver Hader  <oliver@typo3.org>

	* [TASK] #29278: Extend test-case for 1n:FF copy actions on workspaces
	* [BUGFIX] Fix unit-test assertions (array_*_key vs. array_*_assoc)
	* [TASK] #29278: Extend test-case for copying 1n:CSV & 1n:FF to different page on workspaces

2013-02-04  Oliver Hader  <oliver@typo3.org>

	* [TASK] #45132 & #45133: Extend test-case concerning placeholder translation values in workspaces
	* [TASK] #29278 & #45132 & #45133: Extend test-cases concerning sys_refindex
	* [TASK] Disable shadowColumnsForNewPlaceholders in favour of #29278 which solves it implicitly in the Core
	* [TASK] #45151: Extend test-case for 1n:CSV copy actions on workspaces

2013-01-31  Oliver Hader  <oliver@typo3.org>

	* [TASK] Trigger localization of children on parent localization
	* [TASK] #29278: Extend test-case concerning copying and localization of new workspace records
	* [TASK] Remove checks for TYPO3 versions prior to 4.5.0
	* [TASK] #29278: Extend test-case concerning placeholder values in workspaces

2012-11-05  Oliver Hader  <oliver@typo3.org>

	* [TASK] Added FlexForm for IRRE to tt_content records
	* [TASK] Utilize foreign_match_fields to distinguish references

2012-10-12  Oliver Hader  <oliver@typo3.org>

	* [TASK] Added localization in pages_language_overlay of 1nff record

2012-03-12  Oliver Hader  <oliver@typo3.org>

	* Fixed bug: Regular non-workspace tests might be executed on a workspace

2012-03-10  Oliver Hader  <oliver@typo3.org>

	* Cleanup: Changed variables, methods, CGL and install repsonse
	* Release of version 0.4.0

2012-03-06  Oliver Hader  <oliver@typo3.org>

	* Tests: Clean-up test cases
	* Tests: Added test cases for publishing records deleted in workspace

2011-08-12  Oliver Hader  <oliver@typo3.org>

	* Fixed bug: Extensions with hooks in t3lib_TCEmain are not considered to be loaded
	* Fixed bug: Behaviour in TYPO3 4.6 concerning caches has changed and needs special handling

2011-06-20  Oliver Hader  <oliver@typo3.org>

	* Fixed bug: Typing error in install process
	* Tests: Modified test cases to clear TSconfig settings options.workspaces

2011-06-14  Oliver Hader  <oliver@typo3.org>

	* Fixed bug: Calls to deprecated and removed functions

2011-04-05  Oliver Hader  <oliver@typo3.org>

	* Tests: Separated m:n MM localization tests for localizationModes 'keep' and 'select'

2011-03-30  Oliver Hader  <oliver@typo3.org>

	* Tests: Refactored abstract test clases
	* Fixed bug: TCA of m:n MM tables does not use MM relations with uid
	* Tests: Extended m:n MM localization tests

2011-03-23  Oliver Hader  <oliver@typo3.org>

	* Tests: Added m:n MM localization tests
	* Tests: Extended m:n MM localization tests

2011-03-22  Oliver Hader  <oliver@typo3.org>

	* Tests: Refactoring of structure

2011-03-18  Oliver Hader  <oliver@typo3.org>

	* Tests: Added 1:n CSV tests for removing of child records in workspaces
	* Cleanup: Fixed whitespaces in fixtures
	* Tests: Extended 1:n CSV and 1:n FF tests for removing child records in workspaces and flushing dependent elements in the workspace module

2011-03-17  Oliver Hader  <oliver@typo3.org>

	* Fixed bug: Workspaces extension must be loaded if using TYPO3 4.5.0+
	* Tests: Added 1:n FF tests for removing of child records in workspaces

2011-02-09  Oliver Hader  <oliver@typo3.org>

	* Tests: Renamed abstract test class to not be considered as testsuite by PHPUnit
	* Tests: Added skip calls to unsupported test cases that rely on TYPO3 4.5.0

2011-01-21  Oliver Hader  <oliver@typo3.org>

	* Added tabs for forms view for all parent and child tables

2011-01-01  Oliver Hader  <oliver@typo3.org>

	* Tests: Fixed 1:n FF test case since checking on a placeholder for an existing record does not make sense

2010-12-31  Oliver Hader  <oliver@typo3.org>

	* Tests: Extended 1:n FF tests concerning sorting order of placeholder records

2010-12-26  Oliver Hader  <oliver@typo3.org>

	* Tests: Extended 1:n FF tests concerning sorting order of placeholder records

2010-12-23  Oliver Hader  <oliver@typo3.org>

	* Tests: Extended 1:n FF tests concerning child records on pages

2010-11-17  Oliver Hader  <oliver@typo3.org>

	* Release of version 0.3.2

2010-11-13  Oliver Hader  <oliver@typo3.org>

	* Tests: Fixed tests to be working with TYPO3 4.5 and also versions before

2010-11-13  Oliver Hader  <oliver@typo3.org>

	* Tests: Fixed data set of table tx_irretutorial_1nff_price
	* Tests: Extended unit tests for 1:n CSV and 1:n FF
	* Updated sample data to import at ./irre_tutorial/res/T3D__IRRE.t3d
	* Release of version 0.3.1

2010-11-05  Oliver Hader  <oliver@typo3.org>

	* Fixed bug: Inline field for pages table is not available for all doktypes

2010-11-04  Oliver Hader  <oliver@typo3.org>

	* Added parenttable field for prices in 1:n FF context

2010-10-29  Oliver Hader  <oliver@typo3.org>

	* Tests: Fixed labels of error messages to be checked

2010-10-28  Oliver Hader  <oliver@typo3.org>

	* Changed blob fields of 1:n CSV relations to text
	* Updated README.txt contents
	* Tests: Extended unit tests for 1:n FF
	* Tests: Extended unit tests for 1:n CSV

2010-10-27  Oliver Hader  <oliver@typo3.org>

	* Added TCA definition to copy foreign_field and foreign_table_field values to workspace placeholder records

2010-10-20  Oliver Hader  <oliver@typo3.org>

	* Added use case to use 1:n relationships as part of a pages record
	* Updated sample data to import at ./irre_tutorial/res/T3D__IRRE.t3d

2010-10-10  Oliver Hader  <oliver@typo3.org>

	* Added basic skeleton for unit tests
	* Tests: Extended unit tests for 1:n FF

2010-08-06  Oliver Hader  <oliver@typo3.org>

	* Added workspaces and versioning field definitions
	* Raised version number to 0.3.0

2010-02-27  Oliver Hader  <oliver@typo3.org>

	* Release of version 0.2.0

2010-02-27  Oliver Hader  <oliver@typo3.org>


Tests/Unit/Abstract.php
<?php
/***************************************************************
*  Copyright notice
*
*  (c) 2010 Oliver Hader <oliver@typo3.org>
*  All rights reserved
*
*  This script is part of the TYPO3 project. The TYPO3 project is
*  free software; you can redistribute it and/or modify
*  it under the terms of the GNU General Public License as published by
*  the Free Software Foundation; either version 2 of the License, or
*  (at your option) any later version.
*
*  The GNU General Public License can be found at
*  http://www.gnu.org/copyleft/gpl.html.
*
*  This script is distributed in the hope that it will be useful,
*  but WITHOUT ANY WARRANTY; without even the implied warranty of
*  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
*  GNU General Public License for more details.
*
*  This copyright notice MUST APPEAR in all copies of the script!
***************************************************************/

/**
 * Generic test helpers.
 *
 * @author Oliver Hader <oliver@typo3.org>
 */
abstract class tx_datahandlertest_Unit_Abstract extends Tx_Phpunit_Database_TestCase {
	const TABLE_Pages = 'pages';

	const PROPERTY_LocalizeReferencesAtParentLocalization = 'localizeReferencesAtParentLocalization';
	const BEHAVIOUR_LocalizeChildrenAtParentLocalization = 'localizeChildrenAtParentLocalization';
	const BEHAVIOUR_LocalizationMode = 'localizationMode';

	const VALUE_LocalizationMode_Keep = 'keep';
	const VALUE_LocalizationMode_Select = 'select';

	/**
	 * @var string
	 */
	private $path;

	/**
	 * @var integer
	 */
	private $expectedLogEntries = 0;

	/**
	 * @var array
	 */
	private $originalConvVars;

	/**
	 * @var array
	 */
	private $originalTca;

	/**
	 * @var t3lib_beUserAuth
	 */
	private $originalBackendUser;

	/**
	 * @var t3lib_beUserAuth
	 */
	private $backendUser;

	/**
	 * Sets up this test case.
	 *
	 * @return void
	 */
	protected function setUp() {
		$this->expectedLogEntries = 0;

		$this->originalTca = $GLOBALS['TCA'];

		$this->originalBackendUser = clone $GLOBALS['BE_USER'];
		$this->backendUser = $GLOBALS['BE_USER'];
		$this->backendUser->workspace = 0;
		$this->fixBackendUser();

		$this->originalConvVars = $GLOBALS['TYPO3_CONF_VARS'];
		$GLOBALS['TYPO3_CONF_VARS']['SYS']['sqlDebug'] = 1;

		$this->initializeDatabase();
		$this->fixReleatedExtensions();

		$this->flushCaches();
	}

	/**
	 * Tears down this test case.
	 *
	 * @return void
	 */
	protected function tearDown() {
		$this->assertNoLogEntries();

		$this->expectedLogEntries = 0;

		$GLOBALS['TCA'] = $this->originalTca;
		unset($this->originalTca);

		$GLOBALS['TYPO3_CONF_VARS'] = $this->originalConvVars;
		unset($this->originalConvVars);

		$GLOBALS['BE_USER'] = $this->originalBackendUser;
		unset($this->originalBackendUser);
		unset($this->backendUser);

		unset($GLOBALS['T3_VAR']['getUserObj']);

		t3lib_div::purgeInstances();
		$this->dropDatabase();
	}

	/**
	 * Flush all caches that are e.g. used in the workspace module.
	 *
	 * @return void
	 */
	private function flushCaches() {
		$GLOBALS['typo3CacheManager']->flushCaches();
	}

	/**
	 * Since CLI mode does not allow admin user, but setting up ACL for
	 * the unit test is too complex, the cloned user will have admin
	 * permission on the test database.
	 *
	 * @return void
	 */
	private function fixBackendUser() {
		if (defined('TYPO3_REQUESTTYPE_CLI') && TYPO3_REQUESTTYPE_CLI) {
			$this->getBackendUser()->user['admin'] = 1;
		}
	}

	/**
	 * Some extensions register hooks for t3lib_TCEmain that might be executed
	 * during these tests (not a problem) - however the SQL tables have to be
	 * initialized in the test database then.
	 *
	 * @return void
	 */
	private function fixReleatedExtensions() {
		$relatedExtensions = array();
		$hooks =& $GLOBALS['TYPO3_CONF_VARS']['SC_OPTIONS']['t3lib/class.t3lib_tcemain.php'];

		if (is_array($hooks)) {
			foreach ($hooks as $hookExtensions) {
				if (is_array($hookExtensions)) {
					foreach ($hookExtensions as $hookHandler) {
						$matches = array();
						if (is_string($hookHandler) && preg_match('#^EXT:([^/]+)/#', $hookHandler, $matches)) {
							$relatedExtensions[$matches[1]] = TRUE;
						}
					}
				}
			}
		}

		if (count($relatedExtensions)) {
			$this->importExtensions(
				array_keys($relatedExtensions)
			);
		}
	}

	/**
	 * @return boolean
	 */
	protected function createAndUseDatabase() {
		$hasDatabase = $this->createDatabase();

		if ($hasDatabase === TRUE) {
			$this->useTestDatabase();
		} else {
			$this->fail('No test database available');
		}

		return $hasDatabase;
	}

	/**
	 * Initializes a test database.
	 *
	 * @return resource
	 */
	protected function initializeDatabase() {
		$hasDatabase = $this->createAndUseDatabase();

		if ($hasDatabase) {
			$this->importStdDB();
			$this->importExtensions(array('cms'));

			$this->importDataSet($this->getPath() . 'Fixtures/data_pages.xml');
			$this->importDataSet($this->getPath() . 'Fixtures/data_sys_language.xml');
		}

		return $hasDatabase;
	}

	/**
	 * Gets the path to the test directory.
	 *
	 * @return string
	 */
	protected function getPath() {
		if (!isset($this->path)) {
			$this->path = t3lib_extMgm::extPath('datahandler_test') . 'Tests/Unit/';
		}

		return $this->path;
	}

	/**
	 * @return t3lib_beUserAuth
	 */
	protected function getBackendUser() {
		return $this->backendUser;
	}

	/**
	 * Sets the number of expected log entries.
	 *
	 * @param integer $count
	 * @return void
	 */
	protected function setExpectedLogEntries($count) {
		$count = intval($count);

		if ($count > 0) {
			$this->expectedLogEntries = $count;
		}
	}

	/**
	 * Gets the last log entry.
	 *
	 * @return array
	 */
	protected function getLastLogEntryMessage() {
		$message = '';

		$logEntries = $GLOBALS['TYPO3_DB']->exec_SELECTgetRows('*', 'sys_log', 'error IN (1,2)', '', '', 1);

		if (is_array($logEntries) && count($logEntries)) {
			$message = $logEntries[0]['details'];
		}

		return $message;
	}

	/**
	 * @param  array $itemArray
	 * @return array
	 */
	protected function getElementsByItemArray(array $itemArray) {
		$elements = array();

		foreach ($itemArray as $item) {
			$elements[$item['table']][$item['id']] = t3lib_BEfunc::getRecord($item['table'], $item['id']);
		}

		return $elements;
	}

	/**
	 * Gets all records of a table.
	 *
	 * @param string $table Name of the table
	 * @param string $indexField Field of the primary index value (uid)
	 * @param string $orderBy The ORDER BY statement
	 * @return array
	 */
	protected function getAllRecords($table, $indexField = 'uid', $orderBy = '') {
		return $GLOBALS['TYPO3_DB']->exec_SELECTgetRows('*', $table, '1=1', '', $orderBy, '', $indexField);
	}

	/**
	 * Gets the TCE configuration of a field.
	 *
	 * @param  $tableName
	 * @param  $fieldName
	 * @return array
	 */
	protected function getTcaFieldConfiguration($tableName, $fieldName) {
		if (!isset($GLOBALS['TCA'][$tableName]['columns'])) {
			t3lib_div::loadTCA($tableName);
		}

		if (!isset($GLOBALS['TCA'][$tableName]['columns'][$fieldName]['config'])) {
			$this->fail('TCA definition for field ' . $tableName . '.' . $fieldName . ' not available');
		}

		return $GLOBALS['TCA'][$tableName]['columns'][$fieldName]['config'];
	}

	/**
	 * @param string $tableName
	 * @param string $fieldName
	 * @param string $propertyName
	 * @param mixed $value
	 * @return void
	 */
	protected function setTcaFieldConfiguration($tableName, $fieldName, $propertyName, $value) {
		if (!isset($GLOBALS['TCA'][$tableName]['columns'])) {
			t3lib_div::loadTCA($tableName);
		}

		if (isset($GLOBALS['TCA'][$tableName]['columns'][$fieldName]['config'])) {
			$GLOBALS['TCA'][$tableName]['columns'][$fieldName]['config'][$propertyName] = $value;
		}
	}

	/**
	 * @param string $tableName
	 * @param string $fieldName
	 * @param string $behaviourName
	 * @param mixed $value
	 * @return void
	 */
	protected function setTcaFieldConfigurationBehaviour($tableName, $fieldName, $behaviourName, $value) {
		if (!isset($GLOBALS['TCA'][$tableName]['columns'])) {
			t3lib_div::loadTCA($tableName);
		}

		if (isset($GLOBALS['TCA'][$tableName]['columns'][$fieldName]['config'])) {
			if (!isset($GLOBALS['TCA'][$tableName]['columns'][$fieldName]['config']['behaviour'])) {
				$GLOBALS['TCA'][$tableName]['columns'][$fieldName]['config']['behaviour'] = array();
			}

			$GLOBALS['TCA'][$tableName]['columns'][$fieldName]['config']['behaviour'][$behaviourName] = $value;
		}
	}

	/**
	 * Gets the field value of a record.
	 *
	 * @param  $tableName
	 * @param  $id
	 * @param  $fieldName
	 * @return string
	 */
	protected function getFieldValue($tableName, $id, $fieldName) {
		$record = t3lib_BEfunc::getRecord($tableName, $id, $fieldName);

		if (!is_array($record)) {
			$this->fail('Record ' . $tableName . ':' . $id . ' not available');
		}

		return $record[$fieldName];
	}

	/**
	 * Gets instance of t3lib_loadDBGroup.
	 *
	 * @return t3lib_loadDBGroup
	 */
	protected function getLoadDbGroup() {
		$loadDbGroup = t3lib_div::makeInstance('t3lib_loadDBGroup');

		return $loadDbGroup;
	}

	/**
	 * Assert that no sys_log entries had been written.
	 *
	 * @return void
	 */
	protected function assertNoLogEntries() {
		$logEntries = $this->getLogEntries();

		if (count($logEntries) > $this->expectedLogEntries) {
			var_dump(array_values($logEntries)); ob_flush();
			$this->fail('The sys_log table contains unexpected entries.');
		} elseif (count($logEntries) < $this->expectedLogEntries) {
			$this->fail('Expected count of sys_log entries no reached.');
		}
	}

	/**
	 * Asserts the correct order of elements.
	 *
	 * @param string $table
	 * @param string $field
	 * @param array $expectedOrderOfIds
	 * @param string $message
	 * @return void
	 */
	protected function assertSortingOrder($table, $field, $expectedOrderOfIds, $message = NULL) {
		$expectedOrderOfIdsCount = count($expectedOrderOfIds);
		$elements = $this->getAllRecords($table);

		if ($message === NULL) {
			$message = 'Sorting order does not match in table ' . $table;
		}

		foreach ($expectedOrderOfIds as $expectedUid) {
			$this->assertNotEmpty(
				$elements[$expectedUid],
				'Element ' . $table . ':' . $expectedUid . ' not found'
			);
		}

		for ($i = 0; $i < $expectedOrderOfIdsCount-1; $i++) {
			$this->assertLessThan(
				$elements[$expectedOrderOfIds[$i+1]][$field],
				$elements[$expectedOrderOfIds[$i]][$field],
				$message
			);
		}
	}

	/**
	 * Asserts reference index elements.
	 *
	 * @param array $assertions
	 * @param boolean $expected
	 */
	protected function assertReferenceIndex(array $assertions, $expected = TRUE) {
		$references = $this->getAllRecords('sys_refindex', 'hash');

		foreach ($assertions as $parent => $children) {
			foreach ($children as $child) {
				$parentItems = explode(':', $parent);
				$childItems = explode(':', $child);

				$assertion = array(
					'tablename' => $parentItems[0],
					'recuid' => $parentItems[1],
					'field' => $parentItems[2],
					'ref_table' => $childItems[0],
					'ref_uid' => $childItems[1],
				);

				$this->assertTrue(
					($expected === $this->executeAssertionOnElements($assertion, $references)),
					'Expected reference index element for ' . $parent . ' -> ' . $child
				);
			}
		}
	}

	/**
	 * @param string $parentTableName
	 * @param integer $parentId
	 * @param string $parentFieldName
	 * @param array $assertions
	 * @param string $mmTable
	 * @param boolean $expected
	 * @return void
	 */
	protected function assertChildren($parentTableName, $parentId, $parentFieldName, array $assertions, $mmTable = '', $expected = TRUE) {
		$tcaFieldConfiguration = $this->getTcaFieldConfiguration($parentTableName, $parentFieldName);

		$loadDbGroup = $this->getLoadDbGroup();
		$loadDbGroup->start(
			$this->getFieldValue($parentTableName, $parentId, $parentFieldName),
			$tcaFieldConfiguration['foreign_table'],
			$mmTable,
			$parentId,
			$parentTableName,
			$tcaFieldConfiguration
		);

		$elements = $this->getElementsByItemArray($loadDbGroup->itemArray);

		foreach ($assertions as $index => $assertion) {
			$this->assertTrue(
				($expected === $this->executeAssertionOnElements($assertion, $elements)),
				'Assertion #' . $index . ' failed'
			);
		}
	}

	/**
	 * @return array
	 */
	protected function getLogEntries() {
		return $GLOBALS['TYPO3_DB']->exec_SELECTgetRows('*', 'sys_log', 'error IN (1,2)');
	}

	/**
	 * @param  array $assertion
	 * @param  array $elements
	 * @return boolean
	 */
	protected function executeAssertionOnElements(array $assertion, array $elements) {
		if (!empty($assertion['tableName'])) {
			$tableName = $assertion['tableName'];
			unset($assertion['tableName']);
			$elements = (array) $elements[$tableName];
		}

		foreach ($elements as $element) {
			$result = FALSE;

			foreach ($assertion as $field => $value) {
				if ($element[$field] == $value) {
					$result = TRUE;
				} else {
					$result = FALSE;
					break;
				}
			}

			if ($result === TRUE) {
				return TRUE;
			}
		}

		return FALSE;
	}

	/**
	 * @param mixed $element
	 * @return string
	 */
	protected function elementToString($element) {
		$result = preg_replace(
			'#\n+#',
			' ',
			var_export($element, TRUE)
		);

		return $result;
	}

	/**
	 * @return string
	 */
	protected function combine() {
		return implode(':', func_get_args());
	}

	/**
	 * @return tx_datahandlertest_Service_SimulationService
	 */
	protected function getSimulationService() {
		return t3lib_div::makeInstance(
			'tx_datahandlertest_Service_SimulationService'
		);
	}

}

?>

Tests/Unit/AbstractWorkspaces.php
<?php
/***************************************************************
*  Copyright notice
*
*  (c) 2010 Oliver Hader <oliver@typo3.org>
*  All rights reserved
*
*  This script is part of the TYPO3 project. The TYPO3 project is
*  free software; you can redistribute it and/or modify
*  it under the terms of the GNU General Public License as published by
*  the Free Software Foundation; either version 2 of the License, or
*  (at your option) any later version.
*
*  The GNU General Public License can be found at
*  http://www.gnu.org/copyleft/gpl.html.
*
*  This script is distributed in the hope that it will be useful,
*  but WITHOUT ANY WARRANTY; without even the implied warranty of
*  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
*  GNU General Public License for more details.
*
*  This copyright notice MUST APPEAR in all copies of the script!
***************************************************************/

/**
 * Generic test helpers.
 *
 * @author Oliver Hader <oliver@typo3.org>
 */
abstract class tx_datahandlertest_Unit_AbstractWorkspaces extends tx_datahandlertest_Unit_Abstract {
	const VALUE_WorkspaceIdIgnore = -1;

	/**
	 * @var tx_version_tcemain|PHPUnit_Framework_MockObject_MockObject
	 */
	protected $versionTceMainHookMock;

	/**
	 * @var tx_version_tcemain_CommandMap
	 */
	protected $versionTceMainCommandMap;

	/**
	 * Sets up this test case.
	 *
	 * @return void
	 */
	protected function setUp() {
		parent::setUp();

		$this->getBackendUser()->workspace = tx_datahandlertest_Service_SimulationService::WORKSPACE_Draft;
		$this->setWorkspacesConsiderReferences(FALSE);
		$this->setWorkspaceChangeStageMode('');
		$this->setWorkspaceSwapMode('');
	}

	/**
	 * Tears down this test case.
	 *
	 * @return void
	 */
	protected function tearDown() {
		parent::tearDown();

		unset($this->versionTceMainCommandMap);
		unset($this->versionTceMainHookMock);
	}

	/**
	 * Initializes a test database.
	 *
	 * @return resource
	 */
	protected function initializeDatabase() {
		$hasDatabase = parent::initializeDatabase();

		if ($hasDatabase) {
			$this->importExtensions(array('version'));
			$this->importExtensions(array('workspaces'));

			$this->importDataSet($this->getPath() . 'Fixtures/data_sys_workspace.xml');
		}
	}

	/**
	 * Asserts that accordant workspace version exist for live versions.
	 *
	 * @param array $tables Table names with list of ids to be edited
	 * @param integer $workspaceId Workspace to be used
	 * @param boolean $expected The expected value
	 * @return void
	 */
	protected function assertWorkspaceVersions(array $tables, $workspaceId = tx_datahandlertest_Service_SimulationService::WORKSPACE_Draft, $expected = TRUE) {
		foreach ($tables as $tableName => $idList) {
			$ids = t3lib_div::trimExplode(',', $idList, TRUE);
			foreach ($ids as $id) {
				$workspaceVersion = t3lib_BEfunc::getWorkspaceVersionOfRecord($workspaceId, $tableName, $id);
				$this->assertTrue(
					($expected ? $workspaceVersion !== FALSE : $workspaceVersion === FALSE),
					'Workspace version for ' . $tableName . ':' . $id . ($expected ? ' not' : '') . ' availabe'
				);
			}
		}
	}

	/**
	 * Gets a tx_version_tcemain mock.
	 *
	 * @param integer $expectsGetCommandMap (optional) Expects number of invokations to getCommandMap method
	 * @return tx_version_tcemain
	 */
	protected function getVersionTceMainHookMock($expectsGetCommandMap = NULL) {
		$this->versionTceMainHookMock = $this->getMock('tx_version_tcemain', array('getCommandMap'));

		if (is_integer($expectsGetCommandMap) && $expectsGetCommandMap >= 0) {
			$this->versionTceMainHookMock->expects($this->exactly($expectsGetCommandMap))->method('getCommandMap')
				->will($this->returnCallback(array($this, 'getVersionTceMainCommandMapCallback')));
		} elseif (!is_null($expectsGetCommandMap)) {
			$this->fail('Expected invokation of getCommandMap must be integer >= 0.');
		}

		return $this->versionTceMainHookMock;
	}

	/**
	 * Gets access to the command map.
	 *
	 * @param integer $expectsGetCommandMap Expects number of invokations to getCommandMap method
	 * @return void
	 */
	protected function getCommandMapAccess($expectsGetCommandMap) {
		$hookReferenceString = $GLOBALS['TYPO3_CONF_VARS']['SC_OPTIONS']['t3lib/class.t3lib_tcemain.php']['processCmdmapClass']['version'];
		$GLOBALS['T3_VAR']['getUserObj'][$hookReferenceString] = $this->getVersionTceMainHookMock($expectsGetCommandMap);
	}

	/**
	 * @param string $tableName
	 * @param integer $id
	 * @param integer $workspaceId
	 * @param boolean $directLookup
	 * @return boolean
	 */
	protected function getWorkpaceVersionId($tableName, $id, $workspaceId = tx_datahandlertest_Service_SimulationService::WORKSPACE_Draft, $directLookup = FALSE) {
		if ($directLookup) {
			$records = $this->getAllRecords($tableName);
			foreach ($records as $record) {
				if (($workspaceId === self::VALUE_WorkspaceIdIgnore || $record['t3ver_wsid'] == $workspaceId) && $record['t3ver_oid'] == $id) {
					return $record['uid'];
				}
			}
		} else {
			$workspaceVersion = t3lib_BEfunc::getWorkspaceVersionOfRecord($workspaceId, $tableName, $id);
			if ($workspaceVersion !== FALSE) {
				return $workspaceVersion['uid'];
			}
		}

		return FALSE;
	}

	/**
	 * Asserts the existence of a delete placeholder record.
	 *
	 * @param array $tables
	 * @return void
	 */
	protected function assertHasDeletePlaceholder(array $tables) {
		foreach ($tables as $tableName => $idList) {
			$records = $this->getAllRecords($tableName);

			$ids = t3lib_div::trimExplode(',', $idList, TRUE);
			foreach ($ids as $id) {
				$failureMessage = 'Delete placeholder of "' . $tableName . ':' . $id . '"';
				$versionizedId = $this->getWorkpaceVersionId($tableName, $id);
				$this->assertTrue(isset($records[$versionizedId]), $failureMessage . ' does not exist');
				$this->assertEquals($id, $records[$versionizedId]['t3_origuid'], $failureMessage . ' has wrong relation to live workspace');
				$this->assertEquals($id, $records[$versionizedId]['t3ver_oid'], $failureMessage . ' has wrong relation to live workspace');
				$this->assertEquals(2, $records[$versionizedId]['t3ver_state'], $failureMessage . ' is not marked as DELETED');
				$this->assertEquals('DELETED!', $records[$versionizedId]['t3ver_label'], $failureMessage . ' is not marked as DELETED');
			}
		}
	}

	/**
	 * @param array $tables
	 * @return void
	 */
	protected function assertIsDeleted(array $tables) {
		foreach ($tables as $tableName => $idList) {
			$records = $this->getAllRecords($tableName);

			$ids = t3lib_div::trimExplode(',', $idList, TRUE);
			foreach ($ids as $id) {
				$failureMessage = 'Workspaace version "' . $tableName . ':' . $id . '"';
				$this->assertTrue(isset($records[$id]), $failureMessage . ' does not exist');
				$this->assertEquals(0, $records[$id]['t3ver_state']);
				$this->assertEquals(1, $records[$id]['deleted']);
			}
		}
	}

	/**
	 * @param array $tables
	 * @return void
	 */
	protected function assertIsCleared(array $tables) {
		foreach ($tables as $tableName => $idList) {
			$records = $this->getAllRecords($tableName);

			$ids = t3lib_div::trimExplode(',', $idList, TRUE);
			foreach ($ids as $id) {
				$failureMessage = 'Workspaace version "' . $tableName . ':' . $id . '"';
				$this->assertTrue(isset($records[$id]), $failureMessage . ' does not exist');
				$this->assertEquals(0, $records[$id]['t3ver_state'], $failureMessage . ' has wrong state value');
				$this->assertEquals(0, $records[$id]['t3ver_wsid'],  $failureMessage . ' is still in offline workspace');
				$this->assertEquals(-1, $records[$id]['pid'],  $failureMessage . ' has wrong pid value');
			}
		}
	}

	/**
	 * @param array $assertions
	 * @param integer $workspaceId
	 */
	protected function assertRecords(array $assertions, $workspaceId = NULL) {
		foreach ($assertions as $table => $elements) {
			$records = $this->getAllRecords($table);

			foreach ($elements as $uid => $data) {
				$intersection = array_intersect_assoc($data, $records[$uid]);
				$differences = array_intersect_key($records[$uid], array_diff_assoc($data, $records[$uid]));

				$this->assertTrue(
					count($data) === count($intersection),
					'Expected ' . $this->elementToString($data) . ' got differences in ' . $this->elementToString($differences) . ' for table ' . $table
				);

				if (is_integer($workspaceId)) {
					$workspaceVersionId = $this->getWorkpaceVersionId($table, $uid, $workspaceId, TRUE);
					$intersection = array_intersect_assoc($data, $records[$workspaceVersionId]);
					$differences = array_intersect_key($records[$workspaceVersionId], array_diff_assoc($data, $records[$workspaceVersionId]));

					$this->assertTrue(
						count($data) === count($intersection),
						'Expected ' . $this->elementToString($data) . ' got differences in ' . $this->elementToString($differences) . ' for table ' . $table
					);
				}
			}
		}
	}

	/**
	 * Sets the User TSconfig property options.workspaces.considerReferences.
	 *
	 * @param boolean $workspacesConsiderReferences
	 * @return void
	 */
	protected function setWorkspacesConsiderReferences($workspacesConsiderReferences = TRUE) {
		$this->getBackendUser()->userTS['options.']['workspaces.']['considerReferences'] = ($workspacesConsiderReferences ? 1 : 0);
	}

	/**
	 * Sets the User TSconfig property options.workspaces.swapMode.
	 *
	 * @param string $workspaceSwapMode
	 * @return void
	 */
	protected function setWorkspaceSwapMode($workspaceSwapMode = 'any') {
		$this->getBackendUser()->userTS['options.']['workspaces.']['swapMode'] = $workspaceSwapMode;
	}

	/**
	 * Sets the User TSconfig property options.workspaces.changeStageMode.
	 *
	 * @param string $workspaceChangeStateMode
	 * @return void
	 */
	protected function setWorkspaceChangeStageMode($workspaceChangeStateMode = 'any') {
		$this->getBackendUser()->userTS['options.']['workspaces.']['changeStageMode'] = $workspaceChangeStateMode;
	}

	public function getVersionTceMainCommandMapCallback(t3lib_TCEmain $tceMain, array $commandMap) {
		$this->versionTceMainCommandMap = t3lib_div::makeInstance('tx_version_tcemain_CommandMap', $this->versionTceMainHookMock, $tceMain, $commandMap);
		return $this->versionTceMainCommandMap;
	}

	/**
	 * @return tx_version_tcemain_CommandMap
	 */
	protected function getCommandMap() {
		return $this->versionTceMainCommandMap;
	}
}

?>

ext_emconf.php
<?php

########################################################################
# Extension Manager/Repository config file for ext "irre_tutorial".
#
# Auto generated 10-03-2012 21:03
#
# Manual updates:
# Only the data in the array - everything else is removed by next
# writing. "version" and "dependencies" must not be touched!
########################################################################

$EM_CONF[$_EXTKEY] = array(
	'title' => 'DataHandler/TCEmain Test',
	'description' => 'DataHandler/TCEmain Test',
	'category' => 'test',
	'shy' => 0,
	'version' => '0.0.0',
	'dependencies' => '',
	'conflicts' => '',
	'priority' => '',
	'loadOrder' => '',
	'module' => '',
	'state' => 'alpha',
	'uploadfolder' => 0,
	'createDirs' => '',
	'modify_tables' => '',
	'clearcacheonload' => 0,
	'lockType' => '',
	'author' => 'Oliver Hader',
	'author_email' => 'oliver@typo3.org',
	'author_company' => '',
	'CGLcompliance' => '',
	'CGLcompliance_note' => '',
	'constraints' => array(
		'depends' => array(
			'typo3' => '4.5.0-0.0.0',
			'workspaces' => '0.0.0-',
			'version' => '0.0.0-',
		),
		'conflicts' => array(
		),
		'suggests' => array(
		),
	),
	'_md5_values_when_last_written' => '',
	'suggests' => array(
	),
);

?>